### PR TITLE
[API] Add new dlpack API

### PR DESCRIPTION
**Pull request by barry-jin**
## Description ##
Add new dlpack API specified in Array API standardizaiton #20501

References:
1. Discussion on synchronization dmlc/dlpack#57
2. New dlpack API in TVM apache/tvm#7993
3. New dlpack API in pytorch pytorch/pytorch#57110

### Essentials ###
- [ ] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [x] `__dlpack__(self, stream=None)` and `__dlpack_device__(self)`
- [x] Add synchronization between consumer and producer

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [centos-cpu, centos-gpu, unix-cpu, unix-gpu, windows-gpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [centos-gpu, unix-cpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [centos-gpu, unix-cpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [centos-gpu, unix-cpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [centos-gpu, unix-cpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-cpu]

File: src/c_api/c_api.cc [hunk 1]
```diff
@@ -3947,7 +3947,25 @@ int MXShallowCopyNDArray(NDArrayHandle src_handle, NDArrayHandle* out) {
...
+}
+
+int MXGetCurrentStream(int device_id, int* stream) {
+  API_BEGIN();
+#if MXNET_USE_CUDA
+  RunContext rctx{Context::GPU(device_id), nullptr, nullptr};
+  mshadow::Stream<gpu>* cur_stream = rctx.get_stream<gpu>();
+  *stream = reinterpret_cast<int64_t>(mshadow::Stream<gpu>::GetStream(cur_stream));
+#else
+  LOG(FATAL) << "Compile with USE_CUDA=1 to have CUDA runtime compilation.";
```
Comment (szha) [inline src/c_api/c_api.cc:3963]: I'm guessing you meant to report some other error here.

File: tests/python/array-api/test_data_interchange.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        x = torch.tensor((5,), device='cuda:0', dtype=torch.float64) + 1
+    mx = np.from_dlpack(x)
+    stream = torch.cuda.Stream()
+    with torch.cuda.stream(stream):
+        z = torch.from_dlpack(mx)
+    stream.synchronize()
+    z += 1
+    assert z == x
+
+def test_dlpack_mxnet_torch_mxnet():
```
Comment (szha) [inline tests/python/array-api/test_data_interchange.py:35]: Can you also add tests for the expected errors?

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-cpu, unix-gpu, website]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-cpu, unix-gpu, website, windows-gpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-cpu, unix-gpu, website]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-gpu, website]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by barry-jin** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by szha** [comment]
@barry-jin the PR looks OK. Could you resolve the conflicts?

**Comment by barry-jin** [comment]
@mxnet-bot run ci [miscellaneous]

**Pull request merged**